src/masternodes/icxorder.h

```diff
@@ -0,0 +1,490 @@
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action) {
+        READWRITE(dfchtlcTx);
+        READWRITE(seed);
+    }
+};
+
+class CICXClaimDFCHTLCImplemetation : public CICXClaimDFCHTLC
+{
+public:
+    //! tx related properties
+    uint256 creationTx;
+    int32_t creationHeight; 
+
+    CICXClaimDFCHTLCImplemetation()
+        : CICXClaimDFCHTLC()
+        , creationTx()
+        , creationHeight(-1)
+    {}
+    ~CICXClaimDFCHTLCImplemetation() override = default;
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action) {
+        READWRITEAS(CICXClaimDFCHTLC, *this);
+        READWRITE(creationTx);
+        READWRITE(creationHeight);
+    }
+};
+
+struct CICXClaimDFCHTLCMessage : public CICXClaimDFCHTLC {
+    using CICXClaimDFCHTLC::CICXClaimDFCHTLC;
+
+    ADD_SERIALIZE_METHODS;
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action) {
+        READWRITEAS(CICXClaimDFCHTLC, *this);
+    }
+};
+
+class CICXCloseOrder
+{
+public:
+    //! basic properties
+    uint256 orderTx; //txid of order which will be closed
+
+    CICXCloseOrder()
+        : orderTx(uint256())
+    {}
+    virtual ~CICXCloseOrder() = default;
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action) {
+        READWRITE(orderTx);
+    }
+};
+
+class CICXCloseOrderImplemetation : public CICXCloseOrder
+{
+public:
+    //! tx related properties
+    uint256 creationTx;
+    int32_t creationHeight; 
+
+    CICXCloseOrderImplemetation()
+        : CICXCloseOrder()
+        , creationTx()
+        , creationHeight(-1)
+    {}
+    ~CICXCloseOrderImplemetation() override = default;
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action) {
+        READWRITEAS(CICXCloseOrder, *this);
+        READWRITE(creationTx);
+        READWRITE(creationHeight);
+    }
+};
+
+struct CICXCloseOrderMessage : public CICXCloseOrder {
+    using CICXCloseOrder::CICXCloseOrder;
+
+    ADD_SERIALIZE_METHODS;
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action) {
+        READWRITEAS(CICXCloseOrder, *this);
+    }
+};
+
+class CICXOrderView : public virtual CStorageView {
+public:
+    typedef std::pair<DCT_ID,std::string> AssetPair;
+    typedef std::pair<uint8_t,AssetPair> StatusAsset;
+    typedef std::pair<StatusAsset,uint256> OrderKey;
+
+    typedef std::pair<uint256,uint256> TxidPairKey;
+
+    typedef std::pair<uint32_t,uint256> StatusTxid;
+    typedef std::pair<StatusTxid,uint256> StatusTxidKey;
+
+    typedef std::pair<uint32_t,uint256> StatusKey;
```

Comment:
We should rethink these keys, probably we can optimize them.

Reply:
Maybe there could be optimization as I refactored them several times while developing ICX. But most of them is necessary.